### PR TITLE
Property tree

### PR DESCRIPTION
**Pull request by yhmtsai**
This is a small property tree implementation such that we can abstract the reader from different formats for resource manager #800 

This pr contains two parser - rapidjson and command-line.
We can also based on these to develop the subset of json parser to reduce third-party usage.

There will be another pr containing the PoC from the resource manager with this pr to ensure functionalities are enough.

### CODE REVIEW

File: extension/property_tree/include/property_tree/data.hpp [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+        }
+        return stream;
+    }
+
+private:
+    tag_type tag_;
+    std::string str_;
+    long long int int_;
+    double double_;
+    bool bool_;
```
Comment (upsj) [inline extension/property_tree/include/property_tree/data.hpp:144]: a tagged union implementation would look something like
```suggestion
    union {
        long long int int_;
        double double_;
        bool bool_;
    }
```
with placement-new and calling the destructor explicitly, you could also move the `str_` member inside the `union`

File: extension/property_tree/include/property_tree/data.hpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    data_s(float dd) : data_s(static_cast<double>(dd)) {}
+
+    /**
+     * Overloading stream operator<<
+     */
+    friend std::ostream& operator<<(std::ostream& stream, const data_s& data)
+    {
+        switch (data.tag_) {
+        case tag_type::empty_t:
+            stream << "empty";
```
Comment (upsj) [inline extension/property_tree/include/property_tree/data.hpp:121]: nit: something to make clearer we are dealing with an empty object and not a string
```suggestion
            stream << "<empty>";
```

File: extension/property_tree/include/property_tree/data.hpp [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            break;
+        case tag_type::bool_t:
+            stream << (data.bool_ ? "true" : "false");
+            break;
+        case tag_type::int_t:
+            stream << data.int_;
+            break;
+        case tag_type::str_t:
+            stream << data.str_;
```
Comment (upsj) [inline extension/property_tree/include/property_tree/data.hpp:130]: maybe make clearer what's a string?
```suggestion
            stream << '"' << data.str_ << '"';
```

File: extension/property_tree/include/property_tree/data.hpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+     * Constructor for integer
+     */
+    data_s(long long int ii) : tag_(tag_type::int_t) { int_ = ii; }
+
+    /**
+     * Constructor for integer with all integer type
+     */
+    template <typename T, typename = typename std::enable_if<
+                              std::is_integral<T>::value>::type>
+    data_s(T ii) : data_s(static_cast<long long int>(ii))
```
Comment (upsj) [inline extension/property_tree/include/property_tree/data.hpp:91]: nit: this overflows for `uint64_t`

File: extension/property_tree/include/property_tree/property_tree.hpp [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        int size = key_map_.size();
+        auto p = key_map_.emplace(name, size);
+        if (!p.second) {
+            throw std::runtime_error("Have the same key");
+        }
+        children_.emplace_back(pnode_s{name});
+    }
+
+    std::string name_;
+    std::deque<pnode_s<data_type>> children_;
```
Comment (upsj) [inline extension/property_tree/include/property_tree/property_tree.hpp:215]: you only seem to be pushing to the back, and accessing by index, that works better in a vector. Why do we need to use data_type as a template parameter? Can't we set it to `data_type` globally?
```suggestion
    std::vector<pnode_s<data_type>> children_;
```

**Review by upsj — COMMENTED** [review]
a few notes without going through the tests. Really nice approach, I wonder if (since this is an extension) we can just require C++17 support and use `std::variant` instead of reimplementing a tagged union

**Review by MarcelKoch — COMMENTED** [review]
I'm not sure if I understand what this is for. To me, the `property_tree` seems to behave like a thin wrapper for a rapidjson `Document`. I think you could just use that, and store the parsed CLI options also in a rapidjson object.

Also, since we are already relying on third-party for JSON parsing, maybe we can also use third-party tools for CLI options. I personally liked cxxopts.

**Comment by yhmtsai** [comment]
@MarcelKoch Yes, it is a wrapper to accept the rapidjson currently. In the discussion of #800 , there's comment about whether we need to rely on the third-party because it will require all application using this dependence. After this interface, we can provide our own "limited" version for json or other input to avoid leaking dependence. This property_tree can also be done by boost::property_tree, but I implement the one due to the same reason.

**Comment by yhmtsai** [comment]
@upsj yes, it is variant. Not sure about the standard. It's a relatively simple implementation, so may not need to bump the standard version for extension usage
@MarcelKoch thanks for suggestion, does cxxopts support string input? It supports unreconginzed input, so it should be usable.

**Comment by MarcelKoch** [comment]
I'm not entirely sure about the dependency issue, but couldn't that be prevented, if we are using header-only libraries? cxxopts is header only, and I think there are enough good JSON header-only libraries.

**Comment by upsj** [comment]
@MarcelKoch RapidJSON is also header-only, but Ginkgo (without tests etc.) should always be buildable without an internet connection and thus without pulling in dependencies from outside. But as long as this stays an optional component, I think we'd be fine. My vote would be for nlohmann-json, btw, since it is more C++11-y instead of the move-on-copy approach RapidJSON takes since it was started pre C++11 (I think)

**Comment by MarcelKoch** [comment]
@upsj I would also prefer nlohmann-json. Maybe it could be possible to bundle these dependencies for a special offline build, but I agree that disabling the extensions might be the easier choice.

**Comment by upsj** [comment]
One consideration in the past was that other projects use different config file formats (e.g. INI files) that we might want to be able to accomodate, too. That's why I initially suggested using Boost.PropertyTree, but that may be overkill

**Comment by yhmtsai** [comment]
Different file format can always be converted to JSON under some restriction, which is the way I used for resource manager before this pull request. Maybe some license issue from third-party? After pull request, I will just use this one to keep no depedence possible.

**Comment by MarcelKoch** [comment]
nlohman-json and cxxopts are both MIT license, so there should be no issue.

**Comment by keldu** [comment]
Just a minor note. I do separate Formats and the description of the data in my lib.  

The format is mostly some sort of encoding or just Codec.  
The description for the data is called Schema.  

The advantage of using that approach is that you just can do something like.

```
using Schema = Struct < 
  Member < UInt32, "foo" >,
  Member < Float32, "bar" >
>;

// Some c++ like pseudo code
Buffer buffer = read_from_file

JsonCodec json;
Schema::NativeData read_from_json_data = json.template decode <Schema>(buffer);

XmlCodec xml;
Schema::NativeData read_from_xml_data = xml.template decode <Schema>(buffer);
```

You shouldn't use my stuff for it though.
See things like https://capnproto.org/language.html .

The reason why you can't use my stuff is because subtypes aren't supported since I mostly would interpret the types like this.
But capn'proto supports subtypes.
```
struct Node {
  union {
    Node node;
    int leaf;
  } variant;
};
```

**Comment by yhmtsai** [comment]
We tend to rely on the third-party json implementation but more C++-favor one (not rapidjson)

**Pull request closed without merging**